src/util/bip152.rs

```diff
@@ -0,0 +1,335 @@
+// Rust Bitcoin Library
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! BIP152 Compact Blocks
+//!
+//! Implementation of compact blocks data structure and algorithms.
+//!
+
+use std::{error, fmt, io};
+
+use hashes::{hex, sha256d, siphash24};
+
+use blockdata::block::{Block, BlockHeader};
+use blockdata::transaction::Transaction;
+use consensus::encode::{self, Decodable, Encodable, VarInt};
+use util::hash::BitcoinHash;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    /// An unknown version number was used.
+    UnknownVersion,
+    /// The prefill vector has an incorrect length.
+    IncorrectPrefillLength,
+    /// A transaction index is requested that is out
+    /// of range from the corresponding block.
+    TxIndexOutOfRange(u64),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let desc = error::Error::description(self);
+        match *self {
+            Error::TxIndexOutOfRange(i) => write!(f, "{}: {}", desc, i),
+            _ => f.write_str(desc),
+        }
+    }
+}
+
+impl error::Error for Error {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &str {
+        match *self {
+            Error::UnknownVersion => "an unknown version number was used",
+            Error::IncorrectPrefillLength => "the prefill vector has an incorrect length",
+            Error::TxIndexOutOfRange(_) => {
+                "a transaction index is requested that is out of range from the corresponding block"
+            }
+        }
+    }
+}
+
+/// A PrefilledTransaction structure is used in HeaderAndShortIDs to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PrefilledTransaction(pub u64, pub Transaction);
```

Comment:
May be more efficient for users for this to hold a reference to the `Transaction`.

Reply:
This optimization is only possible in the case where you're encoding a compact block, but (1) those are usually transported out of the context so they can't have lifetimes of the original block and (2) by default only the coinbase tx is prefilled and in most cases the nb of prefilled txs will be minimal.